gplugins/design_recipe/DesignRecipe.py

```diff
@@ -84,9 +154,332 @@ def eval_dependencies(self, force_rerun_all=False) -> bool:
+                    results.dirpath = self.dirpath
+                logger.info(
+                    f"Recalled results from {self.dirpath} -> {self.prefix}_results.pkl"
+                )
+        else:
+            with open(str(dirpath.resolve() / f"{self.prefix}_results.pkl"), "rb") as f:
+                unpickler = PathUnpickler(f)
+                results = unpickler.load()
+                if not results.dirpath == dirpath:
+                    results.dirpath = dirpath
+                logger.info(
+                    f"Recalled results from {dirpath} -> {self.prefix}_results.pkl"
+                )
+
+        return results
+
+    def available(self, dirpath: Path | None = None) -> bool:
+        """
+        Check if '*_results.pkl' file exists and results can be loaded
+
+        Parameters:
+            dirpath: Directory with pickle file
+
+        Returns:
+            True if results exist, False otherwise.
+        """
+        if isinstance(dirpath, str):
+            dirpath = Path(dirpath)
+        if dirpath is None:
+            results_file = self.dirpath.resolve() / f"{self.prefix}_results.pkl"
+        else:
+            results_file = dirpath.resolve() / f"{self.prefix}_results.pkl"
+        return results_file.is_file()
+
+
+class Simulation:
+    """
+    Represents the simulation object used to simulate GDSFactory devices.
+
+    This simulation object's purpose is to reduce time simulating by recalling
+    hashed results.
+    """
+
+    # the hash of the system last time convergence was executed
+    last_hash: int = -1
+
+    # A dynamic object used to store convergence results
+    convergence_results: Results
+
+    def __init__(
+        self,
+        component: Component,
+        layerstack: LayerStack | None = None,
+        simulation_settings: pydantic.BaseModel | None = None,
+        convergence_settings: pydantic.BaseModel | None = None,
+        dirpath: Path | None = None,
+    ):
+        self.dirpath = dirpath or Path(".")
+        self.component = component
+        self.layerstack = layerstack or get_layer_stack()
+        self.simulation_settings = simulation_settings
+        self.convergence_settings = convergence_settings
+
+        self.last_hash = hash(self)
+
+        # Create directory for simulation files
+        self.simulation_dirpath = (
+            self.dirpath / f"{self.__class__.__name__}_{self.last_hash}"
+        )
+        self.simulation_dirpath.mkdir(parents=True, exist_ok=True)
+
+        # Create attribute for convergence results
+        self.convergence_results = Results(
+            prefix="convergence", dirpath=self.simulation_dirpath
+        )
+
+    def __hash__(self) -> int:
+        """
+        Returns a hash of all state this Simulation contains
+        Subclasses should include functionality-specific state (e.g. convergence info) here.
+        This is used to determine simulation convergence (i.e. if it needs to be rerun)
```

Comment:
is this used to determine convergence, or is it used to determine freshness?

Reply:
Closer to freshness. It is used to determine whether convergence sweeps have been run already.